### PR TITLE
Add bounds and sanity checks in FV lib

### PR DESCRIPTION
**Pull request by gaojiaqi7**
- Solve the issues found by fuzz
- Add sanity check for firmware volume header

### CODE REVIEW

File: td-shim-tools/src/enroller.rs [hunk 1]
```diff
@@ -32,8 +32,12 @@ pub fn build_cfv_header() -> FvHeader {
         .fv_header
         .file_system_guid
         .copy_from_slice(FIRMWARE_FILE_SYSTEM3_GUID.as_bytes());
+    cfv_header.fv_header.signature = FVH_SIGNATURE;
+    cfv_header.fv_header.header_length = size_of::<FvHeader>() as u16;
     cfv_header.fv_header.fv_length = TD_SHIM_CONFIG_SIZE as u64;
-    cfv_header.fv_header.checksum = 0xdc0a;
+    cfv_header.fv_header.revision = FVH_REVISION;
+    cfv_header.fv_header.update_checksum();
```
Comment (jodh-intel) [inline td-shim-tools/src/enroller.rs:39]: Calculating the checksum looks like an improvement to me but see my comments on this method below.

File: uefi-pi/src/pi/fv.rs [hunk 2]
```diff
@@ -62,6 +76,16 @@ impl FirmwareVolumeHeader {
     pub fn as_bytes(&self) -> &[u8] {
         unsafe { &*slice_from_raw_parts(self as *const Self as *const u8, size_of::<Self>()) }
     }
+
+    // Calculate and update the checksum of the FirmwareVolumeHeader
+    pub fn update_checksum(&mut self) {
+        self.checksum = u16::MAX - sum16(self.as_bytes()) + 1;
+    }
```
Comment (jodh-intel) [inline uefi-pi/src/pi/fv.rs:83]: What worries me is that this method reads the struct as bytes and then modifies that struct to write the checksum back into it. That implies that if there is _already_ a checksum set, the new checksum calculation will be wrong.

A simple fix would be to set `fv_header.checksum = 0` before calling `update_checksum()` I think as that would effectively exclude the checksum from the calculation of... the checksum :smile:

```suggestion
    pub fn update_checksum(&mut self) {
        // Exclude any existing checksum from the checksum calculation
        self.checksum = 0;
        
        self.checksum = u16::MAX - sum16(self.as_bytes()) + 1;
    }
```

That would also make the method idempotent (it currently isn't fwics).

File: uefi-pi/src/pi/fv.rs [hunk 3]
```diff
@@ -266,3 +290,25 @@ impl CommonSectionHeader2 {
         unsafe { &*slice_from_raw_parts(self as *const Self as *const u8, size_of::<Self>()) }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_fvh_checksum() {
```
Comment (jodh-intel) [inline uefi-pi/src/pi/fv.rs:299]: Could you add tests for:

- A `FirmwareVolumeHeader` with an existing checksum to ensure the checksum is ignored.
- Calling `update_checksum()` multiple times to ensure it has no effect on the calculated checksum.

**Review by jodh-intel — COMMENTED** [review]
Thanks @gaojiaqi7. A few comments...

File: uefi-pi/src/pi/fv.rs (hunk 2 above)
Comment (gaojiaqi7) [reply uefi-pi/src/pi/fv.rs]: Right, it was missed, thanks for your suggestion!😊

File: uefi-pi/src/fv.rs [hunk 4]
```diff
@@ -19,15 +19,31 @@ use scroll::Pread;
 
 use crate::pi::fv::*;
 
+// Read FV header from slice and validate its integrity
+pub fn read_fv_header(fv_data: &[u8]) -> Option<FirmwareVolumeHeader> {
```
Comment (jodh-intel) [inline uefi-pi/src/fv.rs:23]: Could you add some unit tests for this function?

Comment (gaojiaqi7) [reply uefi-pi/src/fv.rs:23]: Sure

**Review by jyao1 — APPROVED** [review]

**Pull request merged**